src/librustc_privacy/lib.rs

```diff
@@ -1066,89 +1063,69 @@ impl<'a, 'tcx> SanePrivacyVisitor<'a, 'tcx> {
     /// anything. In theory these qualifiers wouldn't parse, but that may happen
     /// later on down the road...
     fn check_sane_privacy(&self, item: &hir::Item) {
-        let tcx = self.tcx;
-        let check_inherited = |sp: Span, vis: hir::Visibility, note: &str| {
+        let check_inherited = |sp, vis, note: &str| {
             if vis != hir::Inherited {
-                span_err!(tcx.sess, sp, E0449,
-                          "unnecessary visibility qualifier");
+                span_err!(self.tcx.sess, sp, E0449, "unnecessary visibility qualifier");
                 if !note.is_empty() {
-                    tcx.sess.span_note(sp, note);
+                    self.tcx.sess.span_note(sp, note);
                 }
             }
         };
+
         match item.node {
             // implementations of traits don't need visibility qualifiers because
             // that's controlled by having the trait in scope.
             hir::ItemImpl(_, _, _, Some(..), _, ref impl_items) => {
                 check_inherited(item.span, item.vis,
-                                "visibility qualifiers have no effect on trait \
-                                 impls");
+                                "visibility qualifiers have no effect on trait impls");
                 for impl_item in impl_items {
                     check_inherited(impl_item.span, impl_item.vis, "");
                 }
             }
-
-            hir::ItemImpl(..) => {
+            hir::ItemImpl(_, _, _, None, _, _) => {
                 check_inherited(item.span, item.vis,
                                 "place qualifiers on individual methods instead");
             }
+            hir::ItemDefaultImpl(..) => {
+                check_inherited(item.span, item.vis,
+                                "visibility qualifiers have no effect on trait impls");
+            }
             hir::ItemForeignMod(..) => {
                 check_inherited(item.span, item.vis,
-                                "place qualifiers on individual functions \
-                                 instead");
+                                "place qualifiers on individual functions instead");
             }
-
-            hir::ItemEnum(..) |
-            hir::ItemTrait(..) | hir::ItemDefaultImpl(..) |
-            hir::ItemConst(..) | hir::ItemStatic(..) | hir::ItemStruct(..) |
-            hir::ItemFn(..) | hir::ItemMod(..) | hir::ItemTy(..) |
-            hir::ItemExternCrate(_) | hir::ItemUse(_) => {}
```

Comment:
I personally like to always have exhaustive matches wherever possible, e.g. it's easy to skim over this list and say "ah yeah `pub` is allowed on each of these". Perhaps this part could stay?